### ISSUE TITLE
APP_DIRS=False messing with Django Debug Toolbar

### ISSUE THREAD

**Issue by devwaseem** [wrabit/django-cotton#75]
As mentioned here: https://django-debug-toolbar.readthedocs.io/en/latest/installation.html#check-for-prerequisites, DJDT needs APP_DIRS = True, but since django-cotton requires it to be False, the debug toolbar is not rendering correctly. 

is there anyway to fix this?

**Comment by wrabit** [comment]
@devwaseem, I will take a look if something can be done.

**Comment by wrabit** [comment]
You should be able to add django debug toolbar's template directory to the `DIRS` key in your `TEMPLATES` in settings.py as a workaround.

**Comment by Med-ELOMARI** [comment]
First, I’d like to thank you @wrabit of this beautiful package 👏🏻 .

Well @devwaseem I came across the issue too, where the `django_toolbar `templates were being processed by the `cotton_compiler.process` method in the `get_contents `function of the Cotton Loader, which caused some unexpected behavior with the Debug Toolbar.

As a temporary workaround that fits my specific setup, I created a custom loader to bypass processing for particular apps. Here’s the custom loader I used:

``` python
from django.template import TemplateDoesNotExist
from django.template.utils import get_app_template_dirs
from django_cotton.cotton_loader import Loader as CottonLoader

EXCLUDED_APPS = [
    'debug_toolbar'
]

class Loader(CottonLoader):
    @staticmethod
    def get_app_name(origin):
        return origin.template_name.split('/')[0]

    def get_contents(self, origin):
        # Check if the template is from an excluded app
        if self.get_app_name(origin) in EXCLUDED_APPS:
            try:
                # Read the file directly without processing
                with open(origin.name, encoding=self.engine.file_charset) as fp:
                    return fp.read()
            except FileNotFoundError:
                raise TemplateDoesNotExist(origin)
        return super().get_contents(origin)

    def get_dirs(self):
        # Mimic the behavior of "django.template.loaders.filesystem.Loader"
        dirs = self.dirs if self.dirs is not None else self.engine.dirs
        # Mimic the behavior of "django.template.loaders.app_directories.Loader"
        dirs.extend(get_app_template_dirs("templates"))
        return dirs


```
To use this custom loader, I added it to the Django TEMPLATES settings like this:

```python
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [...],
        'APP_DIRS': False,
        'OPTIONS': {
            'loaders': [
                'path.to.your.Loader',
                # no need for filesystem.Loader and app_directories.Loader as get_dirs have their logic
            ],
            ...
        },
    },
]

```
This allows the django_toolbar templates to be used without processing, which resolved the issue for me. However, I'm still trying to understand the exact differences between processed and unprocessed templates, which might be due to the BeautifulSoup parser (html.parser). This is just a workaround and not a scalable solution.

Thanks again for creating such a fantastic package!

**Comment by wrabit** [comment]
Hi @Med-ELOMARI thanks for the response and for the positive remarks on Cotton! 

~What was the issue you had with it in your end?~

Edit: Ok I have resolved it my side, I'll release a fix for it soon.

**Comment by lcfd** [comment]
> Edit: Ok I have resolved it my side, I'll release a fix for it soon.

Thank you @wrabit! I can't wait for the fix.
I think that Cotton deserves more attention from the community, it's the first real answer I found to component-based development ❤️ 

Meanwhile, for all the travelers in search of a temporary solution, the one proposed by @Med-ELOMARI works like a charm, thank you 😊

**Comment by wrabit** [comment]
@lcfd @Med-ELOMARI a patch resolves this in https://github.com/wrabit/django-cotton/pull/80 released as v0.9.26, so @Med-ELOMARI's loader shouldn't be needed now.
